### ISSUE TITLE
"error formatting HCL: illegal char" when string value contains backlash-escaped single quote

### ISSUE THREAD

**Issue by rulatir** [GoogleCloudPlatform/terraformer#420]
### An important warning beforehand

I do NOT have a "minimal reproducible example" because:
  - CloudFormation stacks are HUGE resources (in terms of sheer amount of JSON and HCL).   
  - A "minimal reproducible example" would necessarily have to be something that "otherwise works", i.e. a minimal CloudFormation stack that can actually be spun up, AND containing the specific infrastructural artifact that triggers the terraformer bug I am reporting.
  - Sculpting such a template with clickops would be a tedious, iterative process that would probably take me days or weeks.
  - On the other hand, my description of the symptom alone MIGHT be sufficient to locate the bug relatively easily, i.e. with orders of magnitude less work by terraform developers than would be required of me if I were to undertake this HUGE quest for the holy grail of "minimal reproducible example". I choose to bet on this possibility and only if this proves impossible will I consider going about this the slow and painful way.

### The problem

I am using the `terraformer plan` + `terraformer import plan` workflow. I have a CloudFormation stack whose YAML template defines, among others, an `EcsSpotFleet` resource. This resource contains a string value at the following subpath:
```
.Properties.SpotFleetRequestConfigData.LaunchSpecifications[0].UserData['Fn::Base64']['Fn::Sub']
```
 - that value is a shell script
 - this shell script contains **another** shell script nested in a heredoc block (but this additional nesting level is probably irrelevant)
 - that **inner** shell script contains this line:
```bash
ECS_CLUSTER=\$(curl -s http://localhost:51678/v1/metadata | jq .Cluster | tr -d \")
```
The `tr -d \"` command in the pipeline contains a backslash-escaped quotation mark and this is what trips terraformer's HCL generation.

Specifically, when generating HCL, terraformer seems to escape the backslash, which leaves the quotation mark "naked" and terminating the HCL string.

When reporting this error, terraformer outputs the entirety of the generated HCL to the console. I captured this output and stripped it down to HCL by deleting surrouding lines and stripping line number prefixes. Indeed, the `template_body` property of the `aws_cloudformation_stack` resouce looks like this:
```terraform
   template_body = "<...lots of YAML...>\n       UserData: !<!Base64> \n              'Fn::Sub': >\n              <...lots of shell...>\n              ECS_CLUSTER=\\$(curl -s http://localhost:51678/v1/metadata | jq\n                .Cluster | tr -d \\")\n\n        <...lots of shell...> <...some more YAML...>"
```
As you can see by scrolling the above code viewer to the right, even this code viewer interprets the quote following the escaped backslash as the string delimiter.

### More on why I think the issue might really be quite simple

The `template_body` value should be completely opaque to terraformer. It comes as a single string, it is represented in the plan file JSON as a single string value, and it is supposed to be written into the generated HCL as a single string value. This bug therefore ultimately means that there exist certain string values that terraformer cannot correctly encode as HCL strings. This looks like a rather low-level bug.

**Comment by rulatir** [comment]
EDIT: here are some relevant excerpts from terraformer and terraform files:

`generated/aws/terraformer/eu-central-1/plan.json`:
```
                    "template_body": "... lots of YAML... ECS_CLUSTER=\\$(curl -s http://localhost:51678/v1/metadata | jq .Cluster | tr -d \\\")\n ... more YAML"

It also looks the same in the `terraform.tfstate` file that I generated by manually running `terraform import` to fetch the CF stack that exhibits the issue.

The way the double quote is escaped in those files seems legit. It would seem that it would really be `terraformer` and specifically its HCL renderer that fails.

**Comment by meshuga** [comment]
@rulatir Thanks for raising the problem!
I got a lot of issues with HCL and the way how it performs escaping. Right now it isn't easy, in terraformer we have a set of hacks that make most of the things working.
Personally, I found it much more bullet proof to use JSON output, parameter `--output=json` that prints JSON representation of TF resources [1]. Recently I fixed a bug that was breaking things up when there were `<` or `>` signs in the template [2], so should be working better now (fix hasn't been released yet, you can try out current master). There are still issues with CF interpolation leaking to TF though. No idea how to fix it.

[1] https://www.terraform.io/docs/configuration/syntax-json.html
[2] https://github.com/GoogleCloudPlatform/terraformer/pull/410/files#diff-1bae6ce9b7ae11689ee36ad936acfc48R18-R20

**Comment by rulatir** [comment]
What I currently do is `terraformer plan`, traverse and massage `plan.json` with a tool written in node, then `terraformer import plan`, then un-massage the output. The massaging involves replacing `\"` with `§§§THE_DREADED_ESCAPED_QUOTE§§§` in the `template_body` property of `"aws_cloudformation_stack"` resources, then replacing that magic string with a quote escaped by correct number of backslashes in `.tf`. It is a bit of a horror :fearful:.

For the `import` stage I just prefix an `$` to every `/\${[0-9A-Za-z:]+}/ CF interpolation in plan.json during the same traversal mentioned above. I am not at the `apply` stage yet, so I don't know if this will cause problems, but I don't think so (at least not until you implement that escaping too); I don't believe there is a direct connection between `plan.json` and `terraform.tfstate`, is there? I understand that tfstate is not generated by terraformer but by terraform it runs under the hood?

Regarding solving the CF interpolation issue on terraformer's side, I believe in this case we are importing a string that has no business containing any TF syntax. Therefore if it contains something that **looks like** a TF interpolation, we can just **assume that it isn't**, and escape it with `$$` when writing HCL.

Another thing that gets weirdly escaped is the greater-than sign. If the shell script buried within the template contains a `>`, it ends up as `\>` in HCL, which is an invalid escape sequence. It only happens if template_body is JSON, and weirdly it is solved by parsing and re-stringifying the template during the massage. I may need to run a diff before/after re-stringifying to figure out how what this operation magically changes.

There's also the issue with `"IgnoreKeys"` when an ignored key stores an array or object. I will report this one separately. CF `outputs` is such a key, and it results in `outputs` ending up in HCL, which is invalid because this attribute "cannot be set".

**Comment by meshuga** [comment]
Thanks @rulatir ! Double dollar sign is a correct way to escape dollar signs in HCL and will get to work to fix that :)
Ref https://stackoverflow.com/a/45011180/3784897

**Comment by meshuga** [comment]
When it comes to Terraformer and CF templates - we apply a hack to properly format HCL syntax that can modify the CF template contents and replace occurrences of `= {` to `{`. That has been fixed in #432.
